In [2]:
%matplotlib inline

from pathlib import Path

import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
from mord import LogisticIT
import matplotlib.pylab as plt
import seaborn as sns
from dmba import classificationSummary, gainsChart, liftChart
from dmba.metric import AIC_score

no display found. Using non-interactive Agg backend


In [3]:
sa = pd.read_csv('SystemAdministrators.csv')

In [4]:
sa.head()

Experience  Training Completed task
0        10.9         4            Yes
1         9.9         4            Yes
2        10.4         6            Yes
3        13.7         6            Yes
4         9.4         8            Yes

In [5]:
sa.shape

(75, 3)

In [6]:
sns.scatterplot(x = 'Training', y = 'Experience', hue = 'Completed task', data = sa)

From the scatter plot above, the predictor that appear potentially useful for classifying task completion is Experience.

In [7]:
predictors = ['Experience', 'Training']
outcome = ['Completed task']

In [9]:
X = sa[predictors]
y = sa[outcome]

#partition data
train_X, valid_X, train_y, valid_y = train_test_split(X, y, test_size=0.4, random_state = 1)

#fit a logistic regression (set penalty = l2 and C = 1e42 to avoid regularization)
logit_reg = LogisticRegression(penalty = 'l2', C = 1e42, solver = 'liblinear')
logit_reg.fit(train_X, train_y)



C:\Users\GETHSEMAN\anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


LogisticRegression(C=1e+42, solver='liblinear')

In [10]:
classificationSummary(train_y, logit_reg.predict(train_X))

Confusion Matrix (Accuracy 0.9111)

       Prediction
Actual  0  1
     0 31  1
     1  3 10


The percentage of programmers incorrectly classified(false negative) as failing to complete the task is 1/31 * 100 = 3.22%

In [13]:
print('intercept ', logit_reg.intercept_[0])
print(pd.DataFrame({'coeff': logit_reg.coef_[0]}, index = X.columns).transpose())

intercept  -15.646119443605334
       Experience  Training
coeff    1.835234  0.118728


logit = -15.646 + 1.835E + 0.119T
Probability = 1/1+e^-(logit)
if Prob = 0.5 and T = 4, then E = 8.27

The amount of experience that must be accumulated by a programmer with 4 years of training before his or her estimated probability of completing the task exceed 0.5 is 8.27